File: 0940-fruit-into-baskets/0940-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int n = fruits.size();
        int left = 0;
        int ans = 0;
        unordered_map<int, int> basket;

        for(int right = 0; right < n; right++)
        {
            basket[fruits[right]]++;
            while(basket.size() > 2)
            {
                basket[fruits[left]]--;
                if(basket[fruits[left]] ==0)
                    basket.erase(fruits[left]);
                left++;
            }
            ans = max(ans, right - left + 1);
        }
        return ans;
    }
};

File: 0940-fruit-into-baskets/0940-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        n = len(fruits)
        left = 0
        ans = 0
        basket = {}

        for right in range(n):
            basket[fruits[right]] = basket.get(fruits[right], 0) + 1
            while len(basket) > 2:
                basket[fruits[left]] -= 1
                if basket[fruits[left]] == 0:
                    del basket[fruits[left]]
                left += 1
            ans = max(ans, right - left + 1)
        
        return ans
        